### ISSUE TITLE
Memory leak during back-to-back inferences

### ISSUE THREAD

**Issue by jeremyfowers** [microsoft/onnxruntime-genai#590]
I am experiencing a memory leak while running my application, which is to run an MMLU accuracy test on my Radeon 780M iGPU via DirectML.

Each inference adds tens-hundreds of megabytes to the total system memory and total graphics memory utilized, until it eventually fills up after about 50 inferences and crashes the system.

## My system

* Ryzen "Pheonix" 7940HS with Radeon 780M iGPU
* 32 GB system memory

## Software

* Phi-3-Mini with awq 4 bit weights
* onnxruntime-genai-directml 0.2.0
* onnxruntime-directml 1.18.0

The model is running on the Radeon 780M iGPU,

## My Code

I define a `generate()` function like this, that is meant to return all the response tokens to the input_ids from a prompt.

```python
def generate(
        input_folder,
        input_ids,
        max_new_tokens=512,
        do_sample=True,
        top_k=50,
        top_p=1.0,
        temperature=0.7,
        pad_token_id=None,
    ):
        model = og.Model(input_folder)
        params = og.GeneratorParams(model)

        if pad_token_id:
            params.pad_token_id = pad_token_id

        max_length = len(input_ids) + max_new_tokens

        params.input_ids = input_ids
        params.set_search_options(
            do_sample=do_sample,
            top_k=top_k,
            top_p=top_p,
            temperature=temperature,
            max_length=max_length,
            min_length=max_length,
        )
        params.try_graph_capture_with_max_batch_size(1)

        generator = og.Generator(model, params)

        prompt_start_time = time.perf_counter()
        generator.compute_logits()
        generator.generate_next_token()
        prompt_end_time = time.perf_counter()

        time_to_first_token = prompt_end_time - prompt_start_time

        if max_new_tokens > 1:

            token_gen_times = []
            while not generator.is_done():
                token_gen_start_time = time.perf_counter()
                generator.compute_logits()
                generator.generate_next_token()
                token_gen_end_time = time.perf_counter()

                token_gen_times.append(token_gen_end_time - token_gen_start_time)

            if token_gen_times:
                # List will be empty if we generated 1 or 0 tokens, and we don't
                # want a divide-by-zero error in those cases
                avg_token_gen_latency_s = sum(token_gen_times) / len(
                    token_gen_times
                )
                tokens_per_second = 1 / avg_token_gen_latency_s

        return [generator.get_sequence(0)]
```

Then, I call `generate(tokenizer(prompt), max_new_tokens=1)` dozens of times while running the MMLU accuracy test. Each prompt adds a bit more memory utilization until the system crashes.

## Screenshots

Here is a screenshot of system and iGPU memory utilization. It is climbing like a staircase due to the memory leak, when it should be flat.
![image](https://github.com/microsoft/onnxruntime-genai/assets/80718789/88f6b0e9-4ed4-4be4-8107-dbe6cfcf5feb)
![image](https://github.com/microsoft/onnxruntime-genai/assets/80718789/1a06eb23-6190-4b73-ad73-a9a4444e1a28)

For reference, here is the exact same MMLU accuracy test code running on a Huggingface Transformers implementation of Phi-3-Mini on CPU. Memory utilization is flat, as expected.

![image](https://github.com/microsoft/onnxruntime-genai/assets/80718789/f8a2cff5-c4e3-4221-8fc9-2115265712b1)

## The Question

What do I do about this memory leak? Do I need to do some explicit garbage collection in my code to make my `generate()` function save to run many times in a loop?

**Comment by kunal-vaishnavi** [comment]
Since you are using graph capture, can you try deleting the generator object after an inference run?

https://github.com/microsoft/onnxruntime-genai/blob/8608d1317d9a97101adb3f4fea7889eb348445fb/examples/python/phi3-qa.py#L71-L72

**Comment by kunal-vaishnavi** [comment]
Since you are using graph capture, can you try deleting the generator object after generation is completed?

https://github.com/microsoft/onnxruntime-genai/blob/8608d1317d9a97101adb3f4fea7889eb348445fb/examples/python/phi3-qa.py#L71-L72

**Comment by jeremyfowers** [comment]
I already tried putting del generator and del params at the bottom of my function. I still saw a memory leak.

**Comment by kunal-vaishnavi** [comment]
There has been a recent [ONNX Runtime fix](https://github.com/microsoft/onnxruntime/pull/20879) and a recent [ONNX Runtime GenAI fix](https://github.com/microsoft/onnxruntime-genai/pull/580) for a memory leak issue with DirectML. These fixes will be in the upcoming ONNX Runtime GenAI v0.3.0 release, which is expected to be released this week, and may fix your issue. In the meantime, you can re-build both ONNX Runtime and ONNX Runtime GenAI using the latest commits on the main branches and see if your issue is resolved.

**Comment by jeremyfowers** [comment]
Thanks for the heads up!

**Comment by jeremyfowers** [comment]
@kunal-vaishnavi are there any updates on the 0.3.0 release?

**Comment by kunal-vaishnavi** [comment]
There have been some last-minute PRs that need to be included in the release such as [this one](https://github.com/microsoft/onnxruntime-genai/pull/612). The changes for the v0.3.0 release branch can be tracked [here](https://github.com/microsoft/onnxruntime-genai/pull/602). Once merged, v0.3.0 should be released by end of this week.

**Comment by jeremyfowers** [comment]
@kunal-vaishnavi, very nice meeting you in person the other day!

Today I downloaded 0.3.0 and still saw the memory leak during my MMLU test. So, I decided to dig further and found something interesting.

**The memory leak presented during MMLU, but not during performance benchmarking**. I dug further and found the only meaningful difference between my MMLU and benchmark code was that MMLU delivered a unique prompt on every iteration, whereas my benchmark reused the same prompt across iterations.

Here is a quick psuedocode that has **no** memory leak:

```python
prompt = random_sentence() # generate a sentence of random words with between 100-200 words
for _ in range(1000):
  input_ids = tokenzer.encode(prompt)
  response = model.generate(input_ids)
```

And here is a psuedocode that does show the memory leak:

```python
for _ in range(1000):
  prompt = random_sentence() # generate a sentence of random words with between 100-200 words
  input_ids = tokenzer.encode(prompt)
  response = model.generate(input_ids)
```

The only difference between these two programs is that the plain-text prompt changes between loop iterations.

**Comment by kunal-vaishnavi** [comment]
> @kunal-vaishnavi, very nice meeting you in person the other day!

Very nice to meet you as well!

> The memory leak presented during MMLU, but not during performance benchmarking. I dug further and found the only meaningful difference between my MMLU and benchmark code was that MMLU delivered a unique prompt on every iteration, whereas my benchmark reused the same prompt across iterations.

Thank you for digging further into the memory leak. We will investigate and get back to you.

**Comment by jeremyfowers** [comment]
Thanks @kunal-vaishnavi! Are there any updates?

**Comment by zhangxiang1993** [comment]
Hi @jeremyfowers, we are still investigating it. It takes bit long as it involves a few components.

**Comment by jeremyfowers** [comment]
OK, thank you!

**Comment by contentis** [comment]
+1
Deleting the generator object does not free the activation memory allocated to it, resulting in steadily increasing VRAM when processing a dataset.

**Comment by zhangxiang1993** [comment]
FYI: this bug is much related to a lower level d3d library that directml depends on and is specific to Intel Device, unfortunately, we don't have a quick solution yet, still actively looking into it, but no guranteed ETA yet.

**Comment by jeremyfowers** [comment]
Thanks for your continued efforts on this!

For my education @zhangxiang1993, what do you mean by “Intel Device”? I was running on an all-AMD system.

**Comment by zhangxiang1993** [comment]
Hi @jeremyfowers , sorry for messed up two issues, I meant AMD device for the memory leak issue.

**Comment by jeremyfowers** [comment]
Hey folks, FYI I will be going on parental leave in a few days :)

My team still cares about this issue, though! cc @ramkrishna2910 and @kovtcharov

**Comment by PatriceVignola** [comment]
Hi @jeremyfowers @ramkrishna2910 @kovtcharov

DirectML always uses the graph capture feature for performance reasons and to avoid recompiling graphs on-the-fly, which can be very costly (in that sense, DirectML is closer to TensorRT than CUDA frameworks). This means that when using DirectML, the value of `max_length` in the search options should essentially never change. In your example, `max_length` changes every time you submit a new prompt, which causes DirectML to compile a brand new graph and allocate new memory, which is costly and leads to eventually running out of memory.

Basically, what you need to do is set `max_length` to a value that you think is big enough for your use-case and leave it at that. By keeping `max_length` the same, we can reuse the same graph and memory over and over again. If the limit turned out to be too low, you can delete and recreate the model to free the memory, but it's obviously not ideal. We may add an option in the future to delete previously captured graphs, but it's not something that we can do via the onnxruntime API currently.

**Comment by contentis** [comment]
@PatriceVignola, even if graph rebuilding is required, the allocated memory should be freed up when re-captured. Is this the desired behaviour?

**Comment by PatriceVignola** [comment]
> @PatriceVignola, even if graph rebuilding is required, the allocated memory should be freed up when re-captured. Is this the desired behaviour?

@contentis Not really, but I can see why it would be the assumption. Essentially, when we create a new generator, we capture a graph on-the-fly if there's not already a captured graph with the right sequence length, otherwise we take it from the pool. The thing is, we don't know if future generators will use the same sequence length, but we assume it will since that's the most performant behavior (i.e. reusing generators with the sequence length). So when a generator is destroyed, we returned the graph and its memory to a pool to be picked up by future generators.

If we were to free the captured graph and the memory every time a generator gets destroyed, it wouldn't be very performant. But I can see why some scenarios would require it. What we can do is add an option to disable the pooling of captured graphs and simply get rid of them when generators get destroyed, which would essentially give some memory management control back to the user. That way, the code posted by @jeremyfowers above would work without memory leaks.

**Comment by ramkrishna2910** [comment]
@PatriceVignola yes, that would be ideal. There are cases where we want to generate only 1 additional token (in tests like MMLU). In these cases the `max_length` would change for each call to generate.
It seems that the only way to achieve this right now is to delete and recreate the generator for each prompt.
